Experimental sbt build [ci: last-only]

adriaanm opened the pull request:
Review by @retronym, @lrytz /cc @gkossakowski, @dumpstate.

gkossakowski committed:
Introduce sbt build

It mimics the ant build as closely as necessary to compare the produced
artifacts between the two builds, to build confidence in the transition:

 - all projects are built into the same directories in build/quick
 - include legacy projects: forkjoin, asm, actors
 - TODO: Include forkjoin & asm into library & compiler jars, respectively.

We don't carry over ant's built-in bootstrapping process;
this will be scripted outside of the build, publishing layers locally.

NOTE: the root project cannot be named `scala`:
it shadows the `scala` package in e.g., `projectConsole`.

gkossakowski committed:
Package jars into build/pack/lib directory.

Override default location for jars to be build/pack/lib. This turned out
to be a bit more involved than I expected. Check the diff for details.

Strip down the -version suffix from jar names.

For now we're trying to mimic Ant's build as closely as possible. Ant does
not embed version numbers in jar names it creates in build/pack/lib
directory.

albsadowski committed:
Generate .properties files, pack in jars.

Add an sbt task that generates .properties files that are packaged
into jars. The properties file contains Scala version in various
formats.

The task is registered as a resource generator so sbt can properly
track generated files and make sure they are included in packaged
jars.

The implementation of generateVersionPropertiesFile contains a note
on arbitrary order of written properties. We could consider using
https://github.com/etiennestuder/java-ordered-properties
to guarantee a stable order. Or simply ditch using Properties class
and generate a String. Our property files are simple enough that
nothing more fancy is needed.

gkossakowski committed:
Change project's base to src/$name

Previously, Project.base was set to ./$name and we tweaked manually the
`baseDirectory` setting to point at `.`. This was confusing sbt and its
plugins. We are setting Project.base to ./src/$name and leave
`baseDirectory` intact (so it will point at `./src/$name`.

Adapt settings that depend on `baseDirectory` accordingly.

Refactor configuration as a subproject into a common method.

Given that `baseDirectory` is pointing at `src/$name`, we don't have to
worry about `sourcesInBase`.

gkossakowski committed:
Introduce `buildDirectory` setting.

This settings determines location of where class files, jars, and other
build products will go. By default, it's `./build-sbt`. It makes it easier
to compare what Ant and sbt builds do.

Also, changed `SettingKey` to `settingKey`. The latter is a macro that
automatically determines `label` of a setting and takes a description as
argument. Before that, `SettingKey("my desc")` would create a setting key
with `label` set to "my desc". Not what we wanted.

gkossakowski committed:
Split `commonSettings` into smaller chunks.

Move most of `commonSettings` to `subprojectSettings` and keep
`commonSettings` truly minimal.

gkossakowski committed:
Asm and fork/join: no docs, publishing,...

As they are forks of external Java projects, don't document/publish them.
TODO: move them out of the repo

As in ant, pass -g, -source and -target options to javac to get
byte-identical outputs for Java files between Ant and sbt builds.

gkossakowski committed:
Scaladoc

Put library class files to scaladoc's classpath when documenting the library.
TODO: why is this needed? Bug in scaladoc?

Pass `src/library-aux` to scaladoc using `-doc-no-compile` option.

To mimic ant, output docs into `buildDirectory/scaladoc/$name`.

gkossakowski committed:
Script that diffs `build/` and `build-sbt/`

To ensure the sbt build matches the ant build, diff their output.

The scripts just delegates to `diff` command that compares directories
recursively. We've added some handy ignore patterns (e.g. for .complete
files).

Ignore locker and deps dirs when comparing build and build-sbt,
these directories are needed by Ant build only.

Ignore irrelevant jars in build/pack/lib generated by the Ant build.

gkossakowski committed:
Align how ant builds asm & forkjoin

There were two inconsistencies in building asm & forkjoin:

  1. Asm classes lived in `build/asm`, while forkjoin's were in
     `build/libs/classes/forkjoin`.
  2. Though no jars are needed for these projects, forkjoin was packaged,
     but asm was not.

No reason for these inconsistencies could be found in the history.

gkossakowski committed:
Add "scala-" prefix to the name of a subproject.

Scala artifacts (like library or compiler) are published are published
with "scala-" prefix in their name. We keep project ids short so we can
refer to projects simply by e.g. `library` and not `scala-library`.

Given that `name` settings will have a different value, we switch to
`Project.id` when constructing paths for target directories.

We had to tweak properties generation because over there we use project
name as a file name without "scala-" prefix.

Since our projects are not named uniformly (e.g., "repl" and not "scala-repl"),
add a setting for each subproject that requires special handling.

gkossakowski committed:
Avoid circular deps: override `scalaInstance`

As suggested in sbt/sbt#1872, specify ScalaInstance manually
and use sbt's launcher as a context for Ivy resolution. This way
we avoid the circular dependency problem described in the issue.

gkossakowski committed:
Compiler jar includes interactive, scaladoc & repl

Mirror ant by disabling docs and publishing tasks in these subprojects.
Class files are included by merging `mappings`: the `package` task
uses it to determine which class files should end up in a jar file.

retronym committed:
Exclude transitive scala-library dependency

Scaladoc depends on scala-xml, etc, built with the previous stable
version of Scala. This isn't ideal, but we're stuck with this for
now.

We were relying on Ivy to evict the transitive dependency on the
old version of scala-library. However, SBT treats evictions of
scala-library as suspicious and emits a custom warning.

This commit explicitly excludes the transitive dependency so
as to avoid these warnings. It also makes the output of `show update`
a little easier to read.

gkossakowski committed:
Included required resources in jars

Static resources are stored along source files in our current layout.
Therefore we set resourceDirectory to be sourceDirectory but make sure
that includeFilter includes nothing by default. Subprojects like library
or compiler set their own include filters.

The include filters have been ported from the Ant build definition.

Generate properties for: library, reflect, compiler & actors.

gkossakowski committed:
Root project shouldn't compile any sources.

Let's override `sources` task to return empty collections of sources to
compile. Otherwise, sbt by default would pick up sources in the root
directory which is not what we want.

gkossakowski committed:
Generate shell scripts.

Shell scripts are generated with `mkBin` sbt task that calls ScalaTool.
ScalaTool is defined in project/ and is a port of Ant task that lives
in scala.tools.ant.ScalaTool. While porting, we've simplified the code
significantly.

The `mkBin` task is defined in a new subproject: dist. This subproject
will become responsible for building the distribution in the future.

gkossakowski committed:
Always fork in run.

This way we don't need to worry about mixing classes for Scala compiler
that sbt is invoking and the one we just built. For an example of
possible problem, check out:

https://github.com/gkossakowski/scala/pull/2#issuecomment-78169188

retronym committed:
Projects for partest-extras and junit

 - Explicitly set [un]managed[re]sourceDirectories for all
   projects to improve the import into IntelliJ
 - Add partest extras project, that contains scala/scala specific
   extensions to partest for use in partest and junit tests.
 - Add junit project, which contains unit tests for the compiler and
   library.

retronym committed:
Tweak formatting of build.sbt

gkossakowski committed:
Rework library dependencies

Common library dependencies are extracted into vals.
Those dependencies define "scala-library" as an excluded transitive
dependency so we don't get eviction warnings during `update`.
Lastly, version numbers for dependencies are obtained from
`versions.properties` file.

retronym committed:
`test` task runs Junit tests

Add dependency on scaladoc in junit subproject, as required by
6e6632e (it introduced a junit test for scaladoc).

Enable assertion stack traces and info logging of each test.

retronym committed:
Remove stdout/stderr output from Junit tests

gkossakowski committed:
`test/it:test` runs partest regression tests

Partest tests are defined in Integration configuration.
This means you need to run them with `test/it:test` command. We put them
in separate configuration because they are slow to run so user needs to
ask explicitly for partest to run.

Introduce a common way of setting jar location so it can be reused between
`scalaSubprojectSettings` and `partestJavaAgent`.

gkossakowski committed:
Document sbt build as experimental.

Add welcome message that is shown at sbt's startup. The message clearly
indicates experimental status of sbt build.

Also, add a short paragraph to README.md explaining status of sbt build.

viktorklang commented on build.sbt:
```diff
@@ -0,0 +1,455 @@
+
+  // unfortunately, this will write properties in arbitrary order
+  // this makes it harder to test for stability of generated artifacts
+  // consider using https://github.com/etiennestuder/java-ordered-properties
```
Perhaps Config would be another alternative?

adriaanm replied:
True, were it not that we need to parse them in bash as well :-) Also, no idea who relies on these files since they are part of the jars.

viktorklang replied:
Could be some dragons in there for sure :)

viktorklang commented on build.sbt:
```diff
@@ -0,0 +1,455 @@
+    // uncomment the other definition of the `resolvedArtifactName`
+    val resolvedArtifact = artifact.value
+    val resolvedArtifactName = s"${resolvedArtifact.name}.${resolvedArtifact.extension}"
+    buildDirectory.value / "pack/lib" / resolvedArtifactName
```
Safer with: `buildDirectory.value / "pack" / "lib" / resolvedArtifactName` ?

adriaanm replied:
I was told to concat strings in place as much as possible, as the `/` method doesn't really do much. Never really investigated, to be honest.

viktorklang replied:
Mgith be right, but given that different OS:es uses different path separators I'd leave it to the API to figure it out rather than having it in a string. @jsuereth would know more tho!

adriaanm replied:
As far as I know, when you're generating path names, "/" works equally well on all platforms -- you just can't assume everyone uses it in their paths.

viktorklang commented on build.sbt:
```diff
@@ -0,0 +1,455 @@
+  // we always assume that Java classes are standalone and do not have any dependency
+  // on Scala classes
+  compileOrder := CompileOrder.JavaThenScala,
+  javacOptions in Compile ++= Seq("-g", "-source", "1.5", "-target", "1.6"),
```
Java source encoding?

`"-encoding", "UTF-8"`

adriaanm replied:
The sbt build's first priority is parity with ant -- also, these java sources will disappear from the repo in 2.12, so I don't think we should fiddle with them until then.

fommil commented:
if this works, it will deprecate all the ENSIME setup advice (which might be out of date anyway, because I only submitted the PR to the typelevel fork)

adriaanm commented:
That's not an `if`, as far as I'm concerned -- we absolutely want to make that work seamlessly.

fommil commented:
cool, then please delete the ensime folder (and probably some of the other setup advice files) as they'll all be out of date.

adriaanm commented:
will do -- we still have a ways to go before we get there, though. The sbt build is still experimental for now :)

adriaanm commented:
/nothingtoseehere

retronym committed:
Avoid wildcard ignorance of files named 'target'.

In favour of explicitly ignoring the ones we know contain
SBT build output.

Rationale: we used to have a package named target. Admittedly
we thought better of that (55109d0d253) so it is now called `meta`,
but let's not get lazy and encode a poor practice into our
gitignore.

retronym commented:
I don't think we've got our classpaths straight yet.

https://gist.github.com/6a55ec13b5fa45d04b07

```
% sbt consoleProject
...
scala> currentProject.aggregate.foreach{p => println("=" * 80); Seq(Compile, Test).foreach(conf => println(s"\n= ${p.project}/$conf\n\n" + (dependencyClasspath in (p, conf)).eval.map(_.data.toString).sorted.mkString("\n")))}
...
= interactive/compile

/Users/jason/.ivy2/cache/org.apache.ant/ant-launcher/jars/ant-launcher-1.9.4.jar
/Users/jason/.ivy2/cache/org.apache.ant/ant/jars/ant-1.9.4.jar
/Users/jason/.ivy2/cache/org.scala-lang/scala-library/jars/scala-library-2.11.5.jar
/Users/jason/.ivy2/cache/org.scala-lang/scala-reflect/jars/scala-reflect-2.11.5.jar
/Users/jason/code/scala/build-sbt/libs/classes/asm
/Users/jason/code/scala/build-sbt/libs/classes/forkjoin
/Users/jason/code/scala/build-sbt/quick/classes/compiler
/Users/jason/code/scala/build-sbt/quick/classes/library
/Users/jason/code/scala/build-sbt/quick/classes/reflect
...

= partest-extras/compile

/Users/jason/.ivy2/cache/com.googlecode.java-diff-utils/diffutils/jars/diffutils-1.3.0.jar
/Users/jason/.ivy2/cache/jline/jline/jars/jline-2.12.1.jar
/Users/jason/.ivy2/cache/org.scala-lang.modules/scala-parser-combinators_2.11/bundles/scala-parser-combinators_2.11-1.0.3.jar
/Users/jason/.ivy2/cache/org.scala-lang.modules/scala-partest_2.11/jars/scala-partest_2.11-1.0.6.jar
/Users/jason/.ivy2/cache/org.scala-lang.modules/scala-xml_2.11/bundles/scala-xml_2.11-1.0.3.jar
/Users/jason/.ivy2/cache/org.scala-lang/scala-compiler/jars/scala-compiler-2.11.5.jar
/Users/jason/.ivy2/cache/org.scala-lang/scala-library/jars/scala-library-2.11.5.jar
/Users/jason/.ivy2/cache/org.scala-lang/scala-reflect/jars/scala-reflect-2.11.5.jar
/Users/jason/.ivy2/cache/org.scala-sbt/test-interface/jars/test-interface-1.0.jar
/Users/jason/code/scala/build-sbt/libs/classes/asm
/Users/jason/code/scala/build-sbt/libs/classes/forkjoin
/Users/jason/code/scala/build-sbt/quick/classes/compiler
/Users/jason/code/scala/build-sbt/quick/classes/library
/Users/jason/code/scala/build-sbt/quick/classes/reflect
/Users/jason/code/scala/build-sbt/quick/classes/repl
```

retronym committed:
Avoid unwanted boostrap JARs in the subproject classpaths

I added build configuration to ask SBT to forcible override the Scala
version in the subproject to the declared bootstrap version. I did
this in response to a warning by SBT the declared `scalaVersion`
has been evicted.

However, we actually *want* the newer version. To get rid of the
warning, we need to exclude all `"org.scala-lang" % "*"` when
depending on Scala modules. This cleanly breaks the cycle.

Here's a diff of `*/dependencyClasspath` as a result of this patch:

    https://gist.github.com/retronym/217c76001b1b81798042

retronym commented:
I've got a fix for that problem (which was actually my doing) here: https://github.com/adriaanm/scala/compare/e55bd52...retronym:topic/sbt-tweaks-2?expand=1

I'm no trying out the partest integration, but hit a classloader related problem. Not sure if that is pre-existing, or related to my change.

```
[debug] Framework implementation 'org.specs2.runner.Specs2Framework' not present.
[debug] Framework implementation 'org.specs2.runner.SpecsFramework' not present.
[debug] Framework implementation 'org.specs.runner.SpecsFramework' not present.
[debug] Framework implementation 'org.scalatest.tools.Framework' not present.
[debug] Framework implementation 'org.scalatest.tools.ScalaTestFramework' not present.
[debug] Framework implementation 'com.novocode.junit.JUnitFramework' not present.
[debug] Subclass fingerprints: List((org.scalacheck.Properties,true,sbt.SubclassFingerprintWrapper@5b479ec), (org.scalacheck.Prop,false,sbt.SubclassFingerprintWrapper@2f04211b), (org.scalacheck.Properties,false,sbt.SubclassFingerprintWrapper@495d007), (org.scalacheck.Prop,true,sbt.SubclassFingerprintWrapper@184d548))
[debug] Annotation fingerprints: List((partest,true,scala.tools.partest.Framework$$anon$1@27d9a78d))
java.lang.NoClassDefFoundError: $925ba2fab3293635b76e$$anonfun$test$23$$anonfun$apply$1
    at $925ba2fab3293635b76e$$anonfun$test$23.apply(build.sbt:270)
    at $925ba2fab3293635b76e$$anonfun$test$23.apply(build.sbt:270)
    at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
    at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
    at sbt.std.Transform$$anon$4.work(System.scala:63)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
    at sbt.Execute.work(Execute.scala:235)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
    at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ClassNotFoundException: $925ba2fab3293635b76e$$anonfun$test$23$$anonfun$apply$1
    at scala.tools.nsc.interpreter.AbstractFileClassLoader.findClass(AbstractFileClassLoader.scala:83)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
```

On the bright side, the JUnit test suite passes.

retronym commented:
I can get the partest suite running with: https://gist.github.com/5be77f5c5bcfcc8b9a69. I haven't figured out the root cause (SBT bug? something in my environment?). I think this PR could go ahead while I figure that out.

retronym commented:
`find project -name '*.class' | xargs rm` has make `test/it:test` work for me again without needing 5be77f5c5bcfcc8b9a69.

adriaanm commented:
Thanks! Rebased to include your two commits.

retronym commented:
LGTM

adriaanm commented:
:fireworks:
